twoliter/src/tools.rs

```diff
@@ -0,0 +1,138 @@
+                sentinel_filepath.display(),
+                e
+            );
+            return true;
+        }
+    };
+    trace!("installed hash is '{}'", installed);
+    let do_install = installed != TOOLS_HASH;
+    if do_install {
+        debug!(
+            "Previously installed hash is '{}', our hash '{}'",
+            installed, TOOLS_HASH
+        )
+    }
+    do_install
+}
+
+async fn unpack_tarball(tools_dir: impl AsRef<Path>) -> Result<()> {
+    let tools_dir = tools_dir.as_ref();
+    let tar = GzDecoder::new(TARBALL_DATA);
+    let mut archive = Archive::new(tar);
+    archive.unpack(tools_dir).context(format!(
+        "Unable to unpack tarball into directory '{}'",
+        tools_dir.display()
+    ))?;
+    debug!("Installed tools to '{}'", tools_dir.display());
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_should_install_true_no_sentinel() {
+    let tempdir = tempfile::TempDir::new().unwrap();
+    let sentinel_filepath = tempdir.path().join("installed");
+    assert!(should_install(&sentinel_filepath).await);
+}
+
+#[tokio::test]
+async fn test_should_install_true_different_hashes() {
+    let tempdir = tempfile::TempDir::new().unwrap();
+    let sentinel_filepath = tempdir.path().join("installed");
+    fs::write(&sentinel_filepath, "not the same hash")
+        .await
+        .unwrap();
+    assert!(should_install(&sentinel_filepath).await);
+}
+
+#[tokio::test]
+async fn test_should_install_false() {
+    let tempdir = tempfile::TempDir::new().unwrap();
+    let sentinel_filepath = tempdir.path().join("installed");
+    fs::write(&sentinel_filepath, TOOLS_HASH).await.unwrap();
+    assert!(!should_install(&sentinel_filepath).await);
+}
+
+#[tokio::test]
+async fn test_install_tools() {
+    let tempdir = tempfile::TempDir::new().unwrap();
+    let sentinel_filepath = tempdir.path().join("installed");
+    install_tools(tempdir.path(), false).await.unwrap();
+
+    // Check that the sentinel file has been correctly written.
+    assert_eq!(
+        fs::read_to_string(&sentinel_filepath).await.unwrap(),
+        TOOLS_HASH
+    );
+
+    // Assert that the expected files exist in the tools directory.
+    assert!(tempdir.path().join("Makefile.toml").is_file());
+
+    // Remove a file, then call install_tools again to assert that it did *not* run because it found
+    // matching hashes.
+    std::fs::remove_file(&tempdir.path().join("Makefile.toml")).unwrap();
+    install_tools(tempdir.path(), false).await.unwrap();
+    // The removed file should still *not* be there.
+    assert!(!tempdir.path().join("Makefile.toml").is_file());
+
+    // Call install_tools with force=true and show that it worked by checking that the removed file
+    // is once again found.
+    install_tools(tempdir.path(), true).await.unwrap();
+    // The removed file is there again because we called install_tools with force=true.
+    assert!(tempdir.path().join("Makefile.toml").is_file());
```

Comment:
Can we get rid of the modality here? It seems like running `install_tools()` should always bring the tools into the expected state.

Reply:
It raises a question I have of whether a call to `twoliter install tools` should be required of the user or not. If not, then my goal with to have a one-liner that can be pasted at the top of most commands `install_tools(path, false).await?`.